cpp/osh_bool_stat.h

```diff
@@ -18,26 +18,87 @@ bool isatty(Str* fd_str, word_t* blame_word);
 inline bool DoUnaryOp(Id_t op_id, Str* s) {
   mylib::Str0 path(s);
 
-  // TODO: also call lstat(), etc.
-  struct stat st;
-  if (stat(path.Get(), &st) < 0) {
-    return false;
-  }
-  auto mode = st.st_mode;
-
-  switch (op_id) {
-  // synonyms for existence
-  case Id::BoolUnary_a:
-  case Id::BoolUnary_e:
-    return true;
+  const char *zPath = path.Get();
 
-  case Id::BoolUnary_f:
-    return S_ISREG(mode);
+  if (op_id == Id::BoolUnary_h ||
+      op_id == Id::BoolUnary_L )
+  {
+    struct stat st;
+    if (lstat(zPath, &st) < 0) {
+      return false;
+    }
 
-  case Id::BoolUnary_k:
-    return (mode & S_ISVTX) != 0;
+    return S_ISLNK(st.st_mode);
+  }
+  else
+  {
+    struct stat st;
+    if (stat(zPath, &st) < 0) {
+      return false;
+    }
+
+    auto mode = st.st_mode;
+
+    switch (op_id) {
+    // synonyms for existence
+    case Id::BoolUnary_a:
+    case Id::BoolUnary_e:
+      return true;
+
+    case Id::BoolUnary_s:
+      return st.st_size != 0;
+
+    case Id::BoolUnary_d:
+      return S_ISDIR(mode);
+
+    case Id::BoolUnary_f:
+      return S_ISREG(mode);
+
+    case Id::BoolUnary_k:
+      return (mode & S_ISVTX) != 0;
+
+    case Id::BoolUnary_p:
+      return S_ISFIFO(mode);
+
+    case Id::BoolUnary_O:
+      return st.st_uid == geteuid();
+
+    case Id::BoolUnary_G:
+      return st.st_gid == getegid();
+
+    case Id::BoolUnary_u:
+      return mode & S_ISUID;
+
+    case Id::BoolUnary_g:
+      return mode & S_ISGID;
+
+
+    // NOTE(Jesse): The python OSH interpreter implements -r, -w, -x (in bool_stat.py)
```

Comment:
Alternatively to calling `faccessat()`, `test_st_mode()` does what I outlined in my original comment .. it figures out if the caller has permission to access the file manually using the effective user id, as opposed to the 'real' user id like `access()` does.

Reply:
Ah OK I didn't quite follow the dash code.  But I would suspect that it has some #ifdefs for non-POSIX platforms that are no longer commonly used (e.g. commercial Unix)

I'm not sure what the solution should be but I'd say my preferences are:

1. Avoid configure flags and `#ifdef`, relying instead on portable POSIX code (`faccessat` is POSIX apparently, but older platforms don't necessarily conform to POSIX)
2. Short code (the `test_st_mode()` seems a bit long to copy; it also calls `bash_group_member()`)

But yeah any comments will help, and please file a bug.

Reply:
Another thing I'd say is that if you think you have a solution, and the only blocker is testing as root, it's OK to put "manual" tests in `demo/`.   

That dir is basically a "junk drawer" of stuff that can be moved into tests later (and has been).  There can just be a shell script that runs as root and tests OSH against dash and bash.

But maybe it's better to put that in a different PR (or do it much later) so we don't block this one?  Up to you

Reply:
Well I was unable to repro the bug DASH and CPython work around, so I'm voting to just move on with our lives.  I did open a bug and put a pretty big comment in the code in case anyone ever complains about it.